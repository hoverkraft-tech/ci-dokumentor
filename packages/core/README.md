# core

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build core` to build the library.

## Running unit tests

Run the repository test target which uses Vitest. The canonical command is managed by Nx and the workspace scripts; use the workspace-level test runner so all packages are tested consistently.

• Testing information and canonical commands live in the centralized developer testing guide: `../docs/content/developers/testing.md` (use `pnpm test`, `nx test core`, `pnpm test:ci` etc.).
• Run only the core package tests via Nx: `nx test core`.

Test files are located next to implementation files under `packages/core/src` and use the `.spec.ts` suffix (for example `packages/core/src/.../*.spec.ts`). Prefer referencing those spec files rather than duplicating test code in documentation.
