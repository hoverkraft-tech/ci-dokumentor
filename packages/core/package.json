{
  "name": "@ci-dokumentor/core",
  "description": "Core library for CI Dokumentor",
  "version": "0.0.1",
  "private": true,
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "development": "./src/index.ts",
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "default": "./dist/index.js"
    },
    "./tests": {
      "development": "./__tests__/index.ts",
      "types": "./dist/__tests__/index.d.ts",
      "import": "./dist/__tests__/index.js",
      "default": "./dist/__tests__/index.js"
    }
  },
  "dependencies": {
    "diff": "^8.0.2",
    "fast-glob": "^3.3.3",
    "inversify": "^7.10.3",
    "p-limit": "^7.1.1"
  }
}
