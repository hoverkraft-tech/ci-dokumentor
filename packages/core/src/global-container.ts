import { Container as InversifyContainer } from "inversify";
import { initContainer as coreInitContainer } from './container.js';

export type GlobalContainer = InversifyContainer;

let globalContainer: GlobalContainer | null = null;

export function resetGlobalContainer(): void {
    globalContainer = null;
}

/**
 * Initialize a global container that orchestrates all package containers.
 * This provides a single entry point for dependency injection that includes
 * all services from all packages.
 * 
 * This is the primary runtime container initialization that should be used
 * by both runtime packages and tests for consistent dependency injection.
 */
export function initGlobalContainer(): GlobalContainer {
    if (globalContainer) {
        return globalContainer;
    }

    // Start with core container
    globalContainer = coreInitContainer();

    return globalContainer;
}

/**
 * Initialize global container with all available packages.
 * This dynamically loads and initializes all package containers.
 * 
 * @param packages Optional array of package initializers to include
 */
export async function initGlobalContainerWithPackages(
    packages?: Array<(container: GlobalContainer) => GlobalContainer | Promise<GlobalContainer>>
): Promise<GlobalContainer> {
    if (globalContainer && !packages) {
        return globalContainer;
    }

    // Initialize base global container
    globalContainer = initGlobalContainer();

    // If specific packages are provided, use them
    if (packages) {
        for (const packageInit of packages) {
            globalContainer = await packageInit(globalContainer);
        }
        return globalContainer;
    }

    // Otherwise, try to dynamically load common packages
    try {
        // Try to load git repository package
        const { initContainer: gitInitContainer } = await import('@ci-dokumentor/repository-git');
        gitInitContainer(globalContainer);
    } catch (error) {
        // Package not available, skip
    }

    try {
        // Try to load github repository package  
        const { initContainer: githubInitContainer } = await import('@ci-dokumentor/repository-github');
        githubInitContainer(globalContainer);
    } catch (error) {
        // Package not available, skip
    }

    try {
        // Try to load github actions package
        const { initContainer: githubActionsInitContainer } = await import('@ci-dokumentor/cicd-github-actions');
        githubActionsInitContainer(globalContainer);
    } catch (error) {
        // Package not available, skip
    }

    return globalContainer;
}

/**
 * Initialize global container with specific package initializers (synchronous version).
 * This is primarily for tests that need predictable, synchronous initialization.
 * 
 * @param packageInitializers Array of package initializer functions
 */
export function initGlobalContainerWithPackagesSync(
    packageInitializers: Array<(container: GlobalContainer) => GlobalContainer>
): GlobalContainer {
    // Reset and create fresh container for tests
    globalContainer = initGlobalContainer();

    // Initialize packages synchronously
    for (const packageInit of packageInitializers) {
        globalContainer = packageInit(globalContainer);
    }

    return globalContainer;
}

/**
 * Create a test container with all common packages.
 * This is a convenience function for tests that need a fully configured container.
 */
export function createTestContainer(): GlobalContainer {
    try {
        // Try to require packages synchronously (they should be available in test environment)
        const { initContainer: gitInitContainer } = require('@ci-dokumentor/repository-git');
        const { initContainer: githubInitContainer } = require('@ci-dokumentor/repository-github');
        const { initContainer: githubActionsInitContainer } = require('@ci-dokumentor/cicd-github-actions');

        return initGlobalContainerWithPackagesSync([
            gitInitContainer,
            githubInitContainer,
            githubActionsInitContainer
        ]);
    } catch (error) {
        // Fallback to just core container if packages aren't available
        return initGlobalContainer();
    }
}