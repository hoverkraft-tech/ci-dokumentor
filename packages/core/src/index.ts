
export * from './formatter/formatter.adapter.js'
export * from './formatter/formatter.service.js';
export * from './formatter/markdown-formatter.adapter.js';
export * from './generator/generator.adapter.js'
export * from './generator/generator.service.js';
export * from './generator/section-generator.adapter.js';
export * from './output/output.adapter.js';
export * from './output/file-output.adapter.js';
export * from './repository/repository.service.js';
export * from './container.js';

export { GENERATOR_ADAPTER_IDENTIFIER } from './generator/generator.adapter.js';
export { SECTION_GENERATOR_ADAPTER_IDENTIFIER } from './generator/section-generator.adapter.js';
