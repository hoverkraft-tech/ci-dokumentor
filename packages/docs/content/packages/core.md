---
sidebar_position: 1
---

# Core Package

The `@ci-dokumentor/core` package contains the fundamental building blocks and abstractions that power the entire CI Dokumentor ecosystem.

## Overview

The core package provides:

- **Services** - Core business logic and orchestration
- **Adapters** - Interface implementations for various platforms
- **Interfaces** - Contracts for extensibility
- **Dependencies** - Dependency injection container setup

## Key Components

### Services

#### RepositoryService

Manages repository information and platform detection:

```typescript
class RepositoryService {
  getSupportedRepositoryPlatforms(): string[];
  autoDetectRepositoryPlatform(): Promise<string | null>;
  getRepository(): Promise<Repository>;
}
```

**Repository Model:**

```typescript
type Repository = {
  owner: string;
  name: string;
  url: string;
  fullName: string; // owner/name format
  logo?: string;
  license?: {
    name: string;
    spdxId: string | null;
    url: string | null;
  };
};
```

#### GeneratorService

Manages CI/CD platform adapters and documentation generation:

```typescript
class GeneratorService {
  getSupportedCicdPlatforms(): string[];
  getGeneratorAdapterByPlatform(platform: string): GeneratorAdapter | undefined;
  getSupportedSectionsForPlatform(platform: string): string[];
  autoDetectCicdPlatform(source: string): string | null;
  autoDetectCicdAdapter(source: string): GeneratorAdapter | null;
  generateDocumentationForPlatform(
    adapter: GeneratorAdapter,
    source: string,
    output?: string,
  ): Promise<string>;
}
```

#### FormatterService

Handles content formatting and output adapters:

```typescript
class FormatterService {
  getFormatterAdapterForFile(filePath: string): FormatterAdapter;
  // Additional formatting utilities
}
```

### Interfaces

#### RepositoryProvider

Interface for repository platform providers:

```typescript
interface RepositoryProvider {
  getPlatformName(): string;
  supports(): Promise<boolean>;
  getRepository(): Promise<Repository>;
}
```

#### GeneratorAdapter

Interface for CI/CD platform adapters:

```typescript
interface GeneratorAdapter {
  getPlatformName(): string;
  getSupportedSections(): string[];
  supportsSource(source: string): boolean;
  getDocumentationPath(source: string): string;
  generateDocumentation(
    source: string,
    formatterAdapter: FormatterAdapter,
    outputAdapter: OutputAdapter,
  ): Promise<void>;
}
```

#### FormatterAdapter

Interface for content formatters:

```typescript
interface FormatterAdapter {
  // Formatting methods for different content types
}
```

#### OutputAdapter

Interface for output handling:

```typescript
interface OutputAdapter {
  // Output methods for writing generated content
}
```

### Adapters

#### MarkdownFormatterAdapter

Provides Markdown-specific formatting capabilities.

#### FileOutputAdapter

Handles file system operations for documentation output.

## Dependency Injection

The core package uses [InversifyJS](https://inversify.io/) for dependency injection. The container is set up in `container.ts` and provides bindings for all services and adapters.

## Usage

The core package is used by other packages in the CI Dokumentor ecosystem:

- **CLI Package** - Uses `RepositoryService` and `GeneratorService` for command execution
- **Repository Packages** - Implement `RepositoryProvider` interface
- **CI/CD Packages** - Implement `GeneratorAdapter` interface

## Building

```bash
# Build the core package
nx build core

# Run tests
nx test core

# Run linting
nx lint core
```

### Testing

Testing for the core package is covered by the centralized developer testing guide: `../developers/testing` (see `packages/docs/content/developers/testing.md`). For package-specific tests, open the spec files under `packages/core/src/**/*.spec.ts`.

## Related Packages

- [CLI Package](./cli) - Command-line interface built on core
- [Repository Git](./repository-git) - Git repository provider implementation
- [Repository GitHub](./repository-github) - GitHub-specific repository provider
- [CI/CD GitHub Actions](./cicd-github-actions) - GitHub Actions manifest parser and generator
