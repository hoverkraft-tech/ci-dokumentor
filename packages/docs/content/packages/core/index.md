---
title: Core Package
description: Fundamental building blocks and abstractions powering the CI Dokumentor ecosystem.
sidebar_position: 1
---

The `@ci-dokumentor/core` package contains the fundamental building blocks and abstractions that power the entire CI Dokumentor ecosystem.

## Overview

The core package provides:

- **Services** - Core business logic and orchestration
- **Adapters** - Interface implementations for various platforms
- **Interfaces** - Contracts for extensibility
- **Dependencies** - Dependency injection container setup

## Key Components

### Services

#### RepositoryService

Manages repository information and platform detection:

```typescript
class RepositoryService {
  getSupportedRepositoryPlatforms(): string[];
  autoDetectRepositoryPlatform(): Promise<string | null>;
  getRepository(): Promise<Repository>;
}
```

**Repository Model:**

```typescript
type Repository = {
  owner: string;
  name: string;
  url: string;
  fullName: string; // owner/name format
  logo?: string;
  license?: {
    name: string;
    spdxId: string | null;
    url: string | null;
  };
};
```

**Note**: The runtime `Repository` type in source also includes optional `contributing` metadata (a small object with a `url` field) used by some generator adapters.

#### GeneratorService

Manages CI/CD platform adapters and documentation generation

#### FormatterService

Handles content formatting and output adapters

### Interfaces

#### RepositoryProvider

Interface for repository platform providers:

```typescript
interface RepositoryProvider<
  Options extends RepositoryOptions = RepositoryOptions,
> {
  getPlatformName(): string;

  /**
   * Priority used during auto-detection. Higher values are checked first.
   */
  getPriority(): number;

  /**
   * Check whether this provider supports the current repository context
   */
  supports(): Promise<boolean>;

  /**
   * Return repository information
   */
  getRepository(): Promise<Repository>;

  /**
   * Optional: provide CLI option descriptors specific to this provider
   */
  getOptions(): RepositoryOptionsDescriptors<Options>;

  /**
   * Optional: apply runtime option values to the provider
   */
  setOptions(options: Partial<Options>): void;
}
```

#### GeneratorAdapter

Interface for CI/CD platform adapters:

```typescript
interface GeneratorAdapter {
  getPlatformName(): string;

  getSupportedSections(): string[];

  supportsSource(source: string): boolean;

  getDocumentationPath(source: string): string;

  generateDocumentation(args: {
    source: string;
    sections: GenerateSectionsOptions;
    rendererAdapter: RendererAdapter;
    repositoryProvider: RepositoryProvider;
  }): Promise<void>;
}
```

#### FormatterAdapter

Interface for content formatters:

```typescript
interface FormatterAdapter {
  supportsLanguage(language: FormatterLanguage): boolean;

  heading(input: ReadableContent, level?: number): ReadableContent;

  center(input: ReadableContent): ReadableContent;

  comment(input: ReadableContent): ReadableContent;

  paragraph(input: ReadableContent): ReadableContent;

  bold(input: ReadableContent): ReadableContent;

  italic(input: ReadableContent): ReadableContent;

  code(input: ReadableContent, language?: ReadableContent): ReadableContent;

  inlineCode(input: ReadableContent): ReadableContent;

  link(text: ReadableContent, url: ReadableContent): ReadableContent;

  image(
    url: ReadableContent,
    altText: ReadableContent,
    options?: { width?: string; align?: string },
  ): ReadableContent;

  table(headers: ReadableContent[], rows: ReadableContent[][]): ReadableContent;

  badge(
    label: ReadableContent,
    message: ReadableContent,
    color?: ReadableContent,
  ): ReadableContent;

  lineBreak(): ReadableContent;
}
```

##### MarkdownFormatterAdapter

Provides Markdown-specific formatting capabilities.

#### RendererAdapter

Interface for rendering/finalization adapters (used by generator adapters to produce output or diffs):

```typescript
interface RendererAdapter {
  initialize(
    destination: string,
    formatterAdapter: FormatterAdapter,
  ): Promise<void>;

  getFormatterAdapter(): FormatterAdapter;

  writeSection(sectionIdentifier: string, data: ReadableContent): Promise<void>;

  /**
   * Finalize rendering and return an optional string (for example a diff)
   */
  finalize(): Promise<string | undefined>;
}
```

##### FileRendererAdapter

Handles file system operations for documentation output.

##### DiffRendererAdapter

Generates diffs between existing and new documentation.

## Dependency Injection

The core package uses [InversifyJS](https://inversify.io/) for dependency injection. The container is set up in `container.ts` and provides bindings for all services and adapters.

## Usage

The core package is used by other packages in the CI Dokumentor ecosystem:

- **CLI Package** - Uses `RepositoryService` and `GeneratorService` for command execution
- **Repository Packages** - Implement `RepositoryProvider` interface
- **CI/CD Packages** - Implement `GeneratorAdapter` interface

## Building

```bash
# Build the core package
nx build core

# Run tests
nx test core

# Run linting
nx lint core
```

### Testing

Testing for the core package is covered by the centralized developer testing guide: `../developers/testing` (see `packages/docs/content/developers/testing.md`). For package-specific tests, open the spec files under `packages/core/src/**/*.spec.ts`.

## Related Packages

- [CLI Package](/packages/cli/) - Command-line interface built on core
- [Repository Git](/packages/repository/git/) - Git repository provider implementation
- [Repository GitHub](/packages/repository/github/) - GitHub-specific repository provider
- [CI/CD GitHub Actions](/packages/cicd/github-actions/) - GitHub Actions manifest parser and generator
