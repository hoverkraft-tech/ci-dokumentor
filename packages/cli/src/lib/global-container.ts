import { Container, initContainer as coreInitContainer } from '@ci-dokumentor/core';
import { initContainer as gitInitContainer } from '@ci-dokumentor/repository-git';
import { initContainer as githubInitContainer } from '@ci-dokumentor/repository-github';
import { initContainer as githubActionsInitContainer } from '@ci-dokumentor/cicd-github-actions';

export type GlobalContainer = Container;

let globalContainer: GlobalContainer | null = null;

export function resetGlobalContainer(): void {
    globalContainer = null;
}

/**
 * Initialize a global container that orchestrates all package containers.
 * This provides a single entry point for dependency injection that includes
 * all services from all packages.
 */
export function initGlobalContainer(): GlobalContainer {
    if (globalContainer) {
        return globalContainer;
    }

    // Start with core container
    globalContainer = coreInitContainer();

    // Initialize repository providers
    gitInitContainer(globalContainer);
    githubInitContainer(globalContainer);

    // Initialize CICD providers  
    githubActionsInitContainer(globalContainer);

    return globalContainer;
}