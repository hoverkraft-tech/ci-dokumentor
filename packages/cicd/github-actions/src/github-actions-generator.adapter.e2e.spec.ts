import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { GitHubActionsGeneratorAdapter } from './github-actions-generator.adapter.js';
import {
  FormatterAdapter,
  FileOutputAdapter,
  MarkdownFormatterAdapter,
  RepositoryService,
} from '@ci-dokumentor/core';
import { initGlobalContainer } from './test/global-container.js';
import mockFs from 'mock-fs';
import { existsSync, readFileSync } from 'fs';
import { Repository } from '@ci-dokumentor/core';

describe('GitHubActionsGeneratorAdapter - Integration Tests', () => {
  let formatterAdapter: FormatterAdapter;
  let gitHubActionsGeneratorAdapter: GitHubActionsGeneratorAdapter;
  let repositoryService: RepositoryService;

  beforeEach(async () => {
    // Use real dependencies from the container
    const container = initGlobalContainer();

    formatterAdapter = container.get(MarkdownFormatterAdapter);
    repositoryService = container.get(RepositoryService);
    gitHubActionsGeneratorAdapter = container.get(GitHubActionsGeneratorAdapter);

    // Mock the repository service to return consistent test data
    const mockRepository: Repository = {
      url: 'https://github.com/test-owner/test-action',
      name: 'test-action',
      owner: 'test-owner',
      fullName: 'test-owner/test-action',
      logo: 'https://example.com/logo.png',
    };

    vi.spyOn(repositoryService, 'getRepository').mockResolvedValue(mockRepository);
  });

  afterEach(() => {
    // Restore real file system
    mockFs.restore();
    // Restore all mocks
    vi.restoreAllMocks();
  });

  describe('generateDocumentation', () => {
    it('should generate complete documentation for a GitHub Action', async () => {
      // Arrange
      const actionYaml = `
name: 'Build and Deploy'
description: 'Builds the application and deploys it to the specified environment'
author: 'CI Team'
branding:
  icon: 'upload-cloud'
  color: 'blue'
inputs:
  environment:
    description: 'Target environment (staging, production)'
    required: true
  version:
    description: 'Version to deploy'
    required: false
    default: 'latest'
  dry-run:
    description: 'Run in dry-run mode'
    required: false
    default: 'false'
  config-file:
    description: 'Path to configuration file'
    required: false
    default: 'config/deploy.yml'
outputs:
  deployment-url:
    description: 'URL of the deployed application'
  build-artifacts:
    description: 'Path to build artifacts'
  deployment-status:
    description: 'Status of the deployment (success/failure)'
runs:
  using: 'node20'
  main: 'dist/index.js'
  pre: 'scripts/pre-deploy.js'
  post: 'scripts/post-deploy.js'
`;

      // Setup mock file with the action YAML content
      mockFs({
        '/test/action.yml': actionYaml,
      });

      // Use real FileOutputAdapter for integration testing
      const outputAdapter = new FileOutputAdapter('/test/README.md', formatterAdapter);

      // Act
      await gitHubActionsGeneratorAdapter.generateDocumentation(
        '/test/action.yml',
        formatterAdapter,
        outputAdapter
      );

      // Assert
      const expectedDocumentationPath = '/test/README.md';
      // Verify that the README.md file was created and has content
      expect(existsSync(expectedDocumentationPath)).toBe(true);

      const generatedContent = readFileSync(expectedDocumentationPath, 'utf-8');
      expect(generatedContent).toBeDefined();
      expect(generatedContent.length).toBeGreaterThan(0);

      // Snapshot test the generated content
      expect(generatedContent).toMatchSnapshot('github-action-documentation');
    });

    it('should generate documentation for a CI/CD workflow', async () => {
      // Arrange
      const workflowYaml = `
name: 'CI/CD Pipeline'

on:
  push:
    branches: [main, develop]
  pull_request:
    branches: [main]
  schedule:
    - cron: '0 2 * * 1'
  workflow_dispatch:
    inputs:
      environment:
        description: 'Deployment environment'
        required: true
        default: 'staging'
        type: choice
        options:
          - staging
          - production

env:
  NODE_VERSION: '18'
  REGISTRY: ghcr.io

jobs:
  test:
    name: 'Run Tests'
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [16, 18, 20]
    steps:
      - name: Checkout code
        uses: actions/checkout@v4
      
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: \${{ matrix.node-version }}
          cache: 'npm'
      
      - name: Install dependencies
        run: npm ci
      
      - name: Run tests
        run: npm test
      
      - name: Upload coverage
        uses: codecov/codecov-action@v3
        with:
          file: ./coverage/lcov.info

  build:
    name: 'Build Application'
    runs-on: ubuntu-latest
    needs: test
    outputs:
      image-tag: \${{ steps.meta.outputs.tags }}
      image-digest: \${{ steps.build.outputs.digest }}
    steps:
      - name: Checkout code
        uses: actions/checkout@v4
      
      - name: Setup Docker Buildx
        uses: docker/setup-buildx-action@v3
      
      - name: Login to Container Registry
        uses: docker/login-action@v3
        with:
          registry: \${{ env.REGISTRY }}
          username: \${{ github.actor }}
          password: \${{ secrets.GITHUB_TOKEN }}
      
      - name: Extract metadata
        id: meta
        uses: docker/metadata-action@v5
        with:
          images: \${{ env.REGISTRY }}/\${{ github.repository }}
      
      - name: Build and push
        id: build
        uses: docker/build-push-action@v5
        with:
          context: .
          push: true
          tags: \${{ steps.meta.outputs.tags }}
          labels: \${{ steps.meta.outputs.labels }}

  deploy:
    name: 'Deploy to Environment'
    runs-on: ubuntu-latest
    needs: build
    if: github.ref == 'refs/heads/main'
    environment:
      name: production
      url: https://app.example.com
    steps:
      - name: Deploy application
        run: |
          echo "Deploying \${{ needs.build.outputs.image-tag }}"
          echo "Image digest: \${{ needs.build.outputs.image-digest }}"
`;

      // Setup mock file with the workflow YAML content
      mockFs({
        '/test/.github/workflows/ci-cd.yml': workflowYaml,
      });

      // Use real FileOutputAdapter for integration testing
      const outputAdapter = new FileOutputAdapter('/test/.github/workflows/ci-cd.md', formatterAdapter);

      // Act
      await gitHubActionsGeneratorAdapter.generateDocumentation(
        '/test/.github/workflows/ci-cd.yml',
        formatterAdapter,
        outputAdapter
      );

      // Assert
      const expectedDocumentationPath = '/test/.github/workflows/ci-cd.md';
      // Verify that the README.md file was created and has content
      expect(existsSync(expectedDocumentationPath)).toBe(true);

      const generatedContent = readFileSync(expectedDocumentationPath, 'utf-8');
      expect(generatedContent).toBeDefined();
      expect(generatedContent.length).toBeGreaterThan(0);

      // Snapshot test the generated content
      expect(generatedContent).toMatchSnapshot('workflow-documentation');
    });
  });

  describe('file support validation', () => {
    it('should correctly identify supported GitHub Action files', () => {
      const actionFiles = [
        'action.yml',
        'action.yaml'
      ];

      actionFiles.forEach(file => {
        expect(gitHubActionsGeneratorAdapter.supportsSource(file)).toBe(true);
      });
    });

    it('should correctly identify supported GitHub Workflow files', () => {
      const workflowFiles = [
        '.github/workflows/ci.yml',
        '.github/workflows/ci.yaml',
        '.github/workflows/deploy.yml',
        '.github/workflows/release.yaml'
      ];

      workflowFiles.forEach(file => {
        expect(gitHubActionsGeneratorAdapter.supportsSource(file)).toBe(true);
      });
    });

    it('should reject unsupported files', () => {
      const unsupportedFiles = [
        'package.json',
        'README.md',
        'config.yml',
        'action.json',
        '.github/dependabot.yml'
      ];

      unsupportedFiles.forEach(file => {
        expect(gitHubActionsGeneratorAdapter.supportsSource(file)).toBe(false);
      });
    });
  });

  describe('documentation path generation', () => {
    it('should generate correct paths for GitHub Actions', () => {
      const testCases = [
        { input: 'action.yml', expected: 'README.md' },
        { input: 'action.yaml', expected: 'README.md' },
        { input: 'subfolder/action.yml', expected: 'subfolder/README.md' }
      ];

      testCases.forEach(({ input, expected }) => {
        expect(gitHubActionsGeneratorAdapter.getDocumentationPath(input)).toBe(expected);
      });
    });

    it('should generate correct paths for GitHub Workflows', () => {
      const testCases = [
        { input: '.github/workflows/ci.yml', expected: '.github/workflows/ci.md' },
        { input: '.github/workflows/deploy.yaml', expected: '.github/workflows/deploy.md' },
        { input: '.github/workflows/release.yml', expected: '.github/workflows/release.md' }
      ];

      testCases.forEach(({ input, expected }) => {
        expect(gitHubActionsGeneratorAdapter.getDocumentationPath(input)).toBe(expected);
      });
    });
  });

  describe('error handling', () => {
    it('should handle malformed YAML files gracefully', async () => {
      // Arrange
      const malformedYaml = `
name: 'Malformed Action'
description: 'This YAML has syntax errors
inputs:
  missing-quotes: this should be quoted
    invalid-indentation: true
runs:
  using: 'node20
  main: index.js
`;

      // Setup mock file with malformed YAML
      mockFs({
        '/test/action.yml': malformedYaml
      });

      // Use real FileOutputAdapter for this test
      const outputAdapter = new FileOutputAdapter('/test/README.md', formatterAdapter);

      // Act & Assert
      await expect(
        gitHubActionsGeneratorAdapter.generateDocumentation(
          '/test/action.yml',
          formatterAdapter,
          outputAdapter
        )
      ).rejects.toThrow();
    });

    it('should handle non-existent files gracefully', async () => {
      // Arrange
      // Setup empty mock file system
      mockFs({});

      // Use real FileOutputAdapter for this test
      const outputAdapter = new FileOutputAdapter('/test/README.md', formatterAdapter);

      // Act & Assert
      await expect(
        gitHubActionsGeneratorAdapter.generateDocumentation(
          '/test/action.yml',
          formatterAdapter,
          outputAdapter
        )
      ).rejects.toThrow();
    });

    it('should handle empty YAML files', async () => {
      // Arrange
      // Setup mock file with empty content
      mockFs({
        '/test/action.yml': ''
      });

      // Use real FileOutputAdapter for this test
      const outputAdapter = new FileOutputAdapter('/test/README.md', formatterAdapter);

      // Act & Assert
      // This should either work with empty content or throw a meaningful error
      try {
        await gitHubActionsGeneratorAdapter.generateDocumentation(
          '/test/action.yml',
          formatterAdapter,
          outputAdapter
        );
        // If it succeeds, that's fine
      } catch (error) {
        // If it fails, the error should be meaningful
        expect(error).toBeInstanceOf(Error);
      }
    });
  });
});
