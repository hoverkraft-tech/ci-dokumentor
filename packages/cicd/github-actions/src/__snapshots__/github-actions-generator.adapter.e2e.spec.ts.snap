// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GitHubActionsGeneratorAdapter - Integration Tests > generateDocumentation > should generate complete documentation for a GitHub Action > github-action-documentation 1`] = `
"<!-- header:start -->
<div align="center">
  # ![Icon](data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0IiBmaWxsPSJub25lIiBzdHJva2U9ImN1cnJlbnRDb2xvciIgc3Ryb2tlLXdpZHRoPSIyIiBzdHJva2UtbGluZWNhcD0icm91bmQiIHN0cm9rZS1saW5lam9pbj0icm91bmQiIGNsYXNzPSJmZWF0aGVyIGZlYXRoZXItdXBsb2FkLWNsb3VkIiBjb2xvcj0iYmx1ZSI+PHBvbHlsaW5lIHBvaW50cz0iMTYgMTYgMTIgMTIgOCAxNiI+PC9wb2x5bGluZT48bGluZSB4MT0iMTIiIHkxPSIxMiIgeDI9IjEyIiB5Mj0iMjEiPjwvbGluZT48cGF0aCBkPSJNMjAuMzkgMTguMzlBNSA1IDAgMCAwIDE4IDloLTEuMjZBOCA4IDAgMSAwIDMgMTYuMyI+PC9wYXRoPjxwb2x5bGluZSBwb2ludHM9IjE2IDE2IDEyIDEyIDggMTYiPjwvcG9seWxpbmU+PC9zdmc+) GitHub Action: Build and Deploy
</div>

<!-- header:end -->
<!-- badges:start -->
[![Marketplace](https://img.shields.io/badge/Marketplace-build--and--deploy-blue?logo=github-actions)](https://github.com/marketplace/actions/build-and-deploy)
[![Release](https://img.shields.io/github/v/release/hoverkraft-tech/ci-dokumentor)](https://github.com/hoverkraft-tech/ci-dokumentor/releases)
[![License](https://img.shields.io/github/license/hoverkraft-tech/ci-dokumentor)](https://img.shields.io/github/license/hoverkraft-tech/ci-dokumentor)
[![Stars](https://img.shields.io/github/stars/hoverkraft-tech/ci-dokumentor?style=social)](https://img.shields.io/github/stars/hoverkraft-tech/ci-dokumentor?style=social)

<!-- badges:end -->
<!-- overview:start -->
## Overview

Builds the application and deploys it to the specified environment

<!-- overview:end -->
<!-- usage:start -->
## Usage

\`\`\`yaml
- uses: hoverkraft-tech/ci-dokumentor/test@sha-version
  with:
    # Target environment (staging, production)
    # This input is required.
    environment: ""

    # Version to deploy
    # Default: \`latest\`
    version: latest

    # Run in dry-run mode
    # Default: \`false\`
    dry-run: "false"

    # Path to configuration file
    # Default: \`config/deploy.yml\`
    config-file: config/deploy.yml

\`\`\`

<!-- usage:end -->
<!-- inputs:start -->
## Inputs

| **Input** | **Description** | **Required** | **Default** |
| --- | --- | --- | --- |
| **\`environment\`** | Target environment (staging, production) | **true** | \`\` |
| **\`version\`** | Version to deploy | **false** | \`latest\` |
| **\`dry-run\`** | Run in dry-run mode | **false** | \`false\` |
| **\`config-file\`** | Path to configuration file | **false** | \`config/deploy.yml\` |

<!-- inputs:end -->
<!-- outputs:start -->
## Outputs

| **Output** | **Description** |
| --- | --- |
| **\`deployment-url\`** | URL of the deployed application |
| **\`build-artifacts\`** | Path to build artifacts |
| **\`deployment-status\`** | Status of the deployment (success/failure) |

<!-- outputs:end -->
<!-- secrets:start -->
<!-- secrets:end -->
<!-- contributing:start -->
<!-- contributing:end -->
<!-- license:start -->
<!-- license:end -->
<!-- generated:start -->
This documentation was automatically generated by [CI Dokumentor](https://github.com/hoverkraft-tech/ci-dokumentor).

<!-- generated:end -->
"
`;

exports[`GitHubActionsGeneratorAdapter - Integration Tests > generateDocumentation > should generate documentation for a CI/CD workflow > workflow-documentation 1`] = `
"<!-- header:start -->
<div align="center">
  # GitHub Workflow: CI/CD Pipeline
</div>

<!-- header:end -->
<!-- badges:start -->
[![Release](https://img.shields.io/github/v/release/hoverkraft-tech/ci-dokumentor)](https://github.com/hoverkraft-tech/ci-dokumentor/releases)
[![License](https://img.shields.io/github/license/hoverkraft-tech/ci-dokumentor)](https://img.shields.io/github/license/hoverkraft-tech/ci-dokumentor)
[![Stars](https://img.shields.io/github/stars/hoverkraft-tech/ci-dokumentor?style=social)](https://img.shields.io/github/stars/hoverkraft-tech/ci-dokumentor?style=social)

<!-- badges:end -->
<!-- overview:start -->
<!-- overview:end -->
<!-- usage:start -->
## Usage

\`\`\`yaml
name: CI/CD Pipeline
on:
  push:
    branches:
      - main
      - develop
  pull_request:
    branches:
      - main
  schedule:
    - cron: 0 2 * * 1
jobs:
  ci-cd.yml:
    uses: hoverkraft-tech/ci-dokumentor/test/.github/workflows/ci-cd.yml@sha-version
    secrets:
      # GitHub token for authentication
      github-token: ""
    with:
      # Deployment environment
      # This input is required.
      # Default: \`staging\`
      environment: staging

\`\`\`

<!-- usage:end -->
<!-- inputs:start -->
## Inputs

### Workflow Call Inputs

| **Input** | **Description** | **Required** | **Type** | **Default** |
| --- | --- | --- | --- | --- |
| **\`environment\`** | Deployment environment | **true** | **string** | \`staging\` |

<!-- inputs:end -->
<!-- outputs:start -->
## Outputs

| **Output** | **Description** |
| --- | --- |
| **\`deployment-url\`** | URL of the deployed application |

<!-- outputs:end -->
<!-- secrets:start -->
## Secrets

| **Secret** | **Description** | **Required** |
| --- | --- | --- |
| **\`github-token\`** | GitHub token for authentication | **false** |

<!-- secrets:end -->
<!-- contributing:start -->
<!-- contributing:end -->
<!-- license:start -->
<!-- license:end -->
<!-- generated:start -->
This documentation was automatically generated by [CI Dokumentor](https://github.com/hoverkraft-tech/ci-dokumentor).

<!-- generated:end -->
"
`;
